### PR TITLE
[LLVMGPU][NFC] Unify hal.interface conversion for ROCM and CUDA

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp [hunk 1]
```diff
@@ -336,36 +336,18 @@ class ConvertIREEConstantOp : public ConvertToLLVMPattern {
 };
 
 /// A pattern to convert hal.interface.workgroup.id/count/size into
-/// corresponding NVVM/ROCDL ops.
-template <typename InterfaceOpTy, typename XOp, typename YOp, typename ZOp>
+/// corresponding GPU ops.
+template <typename InterfaceOpTy, typename NewOp>
```
Comment (antiagainst) [inline iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp:340]: Nit: `NewOpTy` to be consistent with `InterfaceOpTy`?

File: iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp [hunk 2]
```diff
@@ -418,6 +378,14 @@ void populateConvertSharedMemoryAllocOps(RewritePatternSet &patterns) {
   patterns.add<ConvertSharedMemAllocOp>(patterns.getContext());
 }
 
+void populateLowerHALInterfaceOp(RewritePatternSet &patterns) {
+  patterns.insert<HALInterfaceWorkgroupOpsConverter<
+                      IREE::HAL::InterfaceWorkgroupIDOp, gpu::BlockIdOp>,
+                  HALInterfaceWorkgroupOpsConverter<
+                      IREE::HAL::InterfaceWorkgroupCountOp, gpu::BlockDimOp>>(
```
Comment (antiagainst) [inline iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp:385]: It should be GridDim ?

File: iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp]: Done

File: iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp]: huh yeah got mixed up somehow, fixed this part

File: iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp [hunk 3]
```diff
@@ -418,6 +378,16 @@ void populateConvertSharedMemoryAllocOps(RewritePatternSet &patterns) {
   patterns.add<ConvertSharedMemAllocOp>(patterns.getContext());
 }
 
+void populateLowerHALInterfaceOp(RewritePatternSet &patterns) {
+  patterns.insert<HALInterfaceWorkgroupOpsConverter<
+                      IREE::HAL::InterfaceWorkgroupIDOp, gpu::BlockIdOp>,
+                  HALInterfaceWorkgroupOpsConverter<
+                      IREE::HAL::InterfaceWorkgroupSizeOp, gpu::BlockDimOp>,
```
Comment (antiagainst) [inline iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp:385]: Do you really need this? I'm not sure you want to translate `InterfaceWorkgroupSizeOp` into `BlockDimOp`. `InterfaceWOrkgroupSizeOp` is [used as the tile size for workgroups](https://github.com/google/iree/blob/main/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp#L999). Although defined so, it's not used as the number of threads in a block (unless each thread just handles one scalar) right now. It's quite confusing, and this is one thing we should resolve..

Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp]: You're right, I don't need it and it is wrong. This get replaced much before. Removed it.

**Review by antiagainst — APPROVED** [review]

**Pull request merged**